include/nbl/builtin/hlsl/cpp_compat/matrix.h

```diff
@@ -1,35 +1,51 @@
 #ifndef _NBL_BUILTIN_HLSL_MATRIX_INCLUDED_
 #define _NBL_BUILTIN_HLSL_MATRIX_INCLUDED_
 
-#ifdef __cplusplus
+#ifndef __HLSL_VERSION 
+#define GLM_FORCE_SWIZZLE
 #include <glm/glm.hpp>
+#include "glm/detail/_swizzle.hpp"
 #include <stdint.h>
 #endif
 
+#include <nbl/builtin/hlsl/cpp_compat/vector.h>
+
 namespace nbl::hlsl
 {
 
-#ifdef __cplusplus
+#ifndef __HLSL_VERSION 
 template<typename T, uint16_t N, uint16_t M>
 struct matrix final : private glm::mat<N,M,T>
 {
     using Base = glm::mat<N,M,T>;
     using Base::Base;
     using Base::operator=;
+    using Base::operator[];
 
     template<uint16_t X, uint16_t Y> matrix(matrix<T, X, Y> const&) = delete;
 
     matrix(matrix const&) = default;
-    matrix(Base const& base) : Base(base) {}
+    explicit matrix(Base const& base) : Base(base) {}
 
-    Base const& asBase() const { return *reinterpret_cast<const Base*>(this); }
-
-    
-    template<class T, uint16_t K>
-    friend matrix<T, N, K> operator*(matrix const& lhs, matrix<T, M, K> const& rhs){ return rhs.asBase() * lhs.asBase(); }
+    friend matrix operator*(matrix const& lhs, matrix const& rhs)
+    { 
+        matrix re = lhs;
+        for(int i = 0; i < N; ++i)
+            lhs[i] *= rhs[i];
+        return re;
+    }
```

Comment:
I don't want this operator to exist, at all

Reply:
because I cannot override it in HLSL to be "proper" we shouldn't have this in C++ as to discourage its use in HLSL